File: phase2_ekf/ekf.py
class ExtendedKalmanFilter:
    def __init__(self, x0, P0, Q, R, physics_model):
        """
        Args:
            x0: initial state estimate (scalar)
            P0: initial covariance (scalar)
            Q: process noise covariance (scalar)
            R: measurement noise covariance (scalar)  
            physics_model: PhysicsModel instance
        """
        self.x = x0
        self.P = P0
        self.Q = Q
        self.R = R
        self.model = physics_model
        
        self.nis = 0.0
        self.innovation = 0.0
        self.S = 0.0
        self.x_pred = x0
        self.P_pred = P0
    
    def predict(self, u):
        """EKF prediction step.
        x_pred = f(x, u)
        P_pred = F * P * F + Q
        """
        self.x_pred = self.model.state_transition(self.x, u)
        F = self.model.jacobian_F(self.x, u)
        self.P_pred = F * self.P * F + self.Q
    
    def update(self, z):
        """EKF update step.
        innovation y = z - h(x_pred)
        S = H * P_pred * H + R
        K = P_pred * H / S
        x = x_pred + K * y
        P = (1 - K*H) * P_pred
        NIS = y^2 / S
        
        Returns: (x_est, P, nis, innovation)
        """
        H = self.model.jacobian_H(self.x_pred)
        h_x = self.model.measurement_function(self.x_pred)
        
        self.innovation = z - h_x
        self.S = H * self.P_pred * H + self.R
        
        K = self.P_pred * H / self.S
        
        self.x = self.x_pred + K * self.innovation
        self.P = (1.0 - K * H) * self.P_pred
        
        self.nis = (self.innovation ** 2) / self.S
        return self.x, self.P, self.nis, self.innovation
    
    def step(self, u, z):
        """Combined predict + update.
        Returns: (x_est, P, nis, innovation)
        """
        self.predict(u)
        return self.update(z)
    
    def predict_only(self, u):
        """Predict without update (for when gate says SLEEP and we skip measurement update to protect state from corrupted readings)."""
        self.predict(u)
        self.x = self.x_pred
        self.P = self.P_pred
        return self.x, self.P

File: Core/Src/ekf.c
#include "ekf.h"

void EKF_Init(EKF_t *ekf, float x0, float P0) {
    ekf->x = x0;
    ekf->P = P0;
    ekf->Q = EKF_Q_VAR;
    ekf->R = EKF_R_VAR;
}

float EKF_Step(EKF_t *ekf, float z, float pred_physics) {
    // 1. Predict
    // State transition f(x) = (1 - TAU)*x + TAU*pred_physics
    float x_pred = (1.0f - TAU) * ekf->x + TAU * pred_physics;
    
    // Jacobian F = (1 - TAU)
    float F = 1.0f - TAU;
    float P_pred = (F * ekf->P * F) + ekf->Q;
    
    // 2. Update
    // Measurement h(x) = x, so H = 1.0
    float innovation = z - x_pred;
    float S = P_pred + ekf->R;
    
    // Kalman Gain K = P_pred * H / S (where H=1)
    float K = P_pred / S;
    
    // State update
    ekf->x = x_pred + K * innovation;
    
    // Covariance update P = (1 - K*H) * P_pred
    ekf->P = (1.0f - K) * P_pred;
    
    // Calculate and return NIS (Normalized Innovation Squared)
    float nis = (innovation * innovation) / S;
    
    return nis;
}
